### DISCUSSION TITLE
TmuxPane

### DISCUSSION THREAD

**Discussion by Dixeran** [wezterm/wezterm#1807] [Ideas]
Currently we use `LocalPane` to render tmux contents, but we definitely need a `TmuxPane` to get more control about the rendering (most of all the scrollback). 

I'm thinking about wrap a LocalPane inside TmuxPane so that all new contents come after attach will feed into LocalPane, render as normal, like what we've done now. Additionally, we can get a `history_size` from tmux that indicate how many lines are there in history, and reserve empty lines in Screen (by `scroll_up` or something). When user scroll up, we fetch history contents from remote and render into those empty lines (is it possible?).

The potential problem is that after the resize, the tracking of the history becomes difficult.

**Comment by wez** [comment]
I'm just waking up so this is probably a half-baked thought, but: rather than use a `LocalPane`, the `TmuxPane` could build a `Terminal` of its own, and limit that terminal to 0 lines of scrollback.  The main viewport would render from the `Terminal`'s lines, but the `TmuxPane` and its associated `Renderable` would know how to request scrollback from tmux instead of the `Terminal`.

When it is resized, we could "simply" request the full viewport from tmux and apply that to the `Terminal`.

**Reply by MuhammedZakir** [reply]
Can't this be made generic so that similiar apps such as screen or dvtm can also benefit? Scrollback and other needed things could be done via Lua callbacks?

**Reply by wez** [reply]
I don't believe that either screen or dtvm have anything similar to tmux CC mode, and even if they did, I have very little interest in adding support for other multiplexers beyond tmux.

wezterm has its own multiplexer feature that is the preferred way to manage this sort of thing.  Adding lots of different backends will make maintainance of wezterm much more difficult and the result is that each of those backends will be worse off than if we keep things focused.

I'm willing to include tmux support as a way to improve the user experience for systems that cannot install the wezterm multiplexer on the remote host, but it's never going to be as good as wezterm's multiplexer is able to be.

**Reply by Dixeran** [reply]
There might still need another Terminal to handle scrollback contents (VT sequence from tmux)?
